src/enumerable.cr

```diff
@@ -609,6 +609,37 @@ module Enumerable(T)
     memo
   end
 
+  # Similar to `reduce`, but instead of raising when the input is empty,
+  # return `nil`
+  #
+  # ```
+  # ([] of Int32).reduce? { |acc, i| acc + i } # => nil
+  # ```
+  def reduce?
+    memo = uninitialized T
+    found = false
+
+    each do |elem|
+      memo = found ? (yield memo, elem) : elem
+      found = true
+    end
+
+    found ? memo : nil
+  end
+
+  # Similar to `reduce?`, however instead of returning `nil` when the input is empty,
+  # return the initial value of the accumulator.
```

Comment:
Maybe not? I wasn't certain of how a user might write their code, and how this might be used in that regard. I can remove the `reduce?(memo)` if that's the right thing.